File: article/src/main/java/com/ruzhan/awaker/article/source/AwakerRepository.java
package com.ruzhan.awaker.article.source;


import com.ruzhan.awaker.article.App;
import com.ruzhan.awaker.article.db.AwakerArticleAppDatabase;
import com.ruzhan.awaker.article.db.entity.CommentEntity;
import com.ruzhan.awaker.article.db.entity.NewsEntity;
import com.ruzhan.awaker.article.db.entity.SpecialListEntity;
import com.ruzhan.awaker.article.db.entity.UserInfoEntity;
import com.ruzhan.awaker.article.model.BannerItem;
import com.ruzhan.awaker.article.model.Comment;
import com.ruzhan.awaker.article.model.NewDetail;
import com.ruzhan.awaker.article.model.News;
import com.ruzhan.awaker.article.model.Special;
import com.ruzhan.awaker.article.model.SpecialDetail;
import com.ruzhan.awaker.article.model.UserInfo;
import com.ruzhan.awaker.article.network.AwakerClient;
import com.ruzhan.awaker.article.network.HttpResult;
import com.ruzhan.awaker.article.source.remote.IRemoteDataSource;
import com.ruzhan.awaker.article.source.remote.RemoteDataSourceImpl;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;

/**
 * Copyright ©2017 by ruzhan
 */

public final class AwakerRepository implements IRemoteDataSource {

    private static AwakerRepository INSTANCE;

    private IRemoteDataSource remoteDataSource;
    private AwakerArticleAppDatabase awakerArticleAppDatabase;

    private AwakerRepository() {
        remoteDataSource = new RemoteDataSourceImpl(AwakerClient.get());
        awakerArticleAppDatabase = AwakerArticleAppDatabase.get(App.get());
    }

    public static AwakerRepository get() {
        if (INSTANCE == null) {
            synchronized (AwakerRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new AwakerRepository();
                }
            }
        }
        return INSTANCE;
    }

    public Flowable<UserInfoEntity> loadUserInfoEntity(String id) {
        return awakerArticleAppDatabase.userInfoDao().loadUserInfoEntity(id);
    }

    public void insertUserInfoEntity(UserInfoEntity userInfoEntity) {
        awakerArticleAppDatabase.userInfoDao().insertUserInfoEntity(userInfoEntity);
    }

    public Flowable<SpecialDetail> loadSpecialDetail(String id) {
        return awakerArticleAppDatabase.specialDetailDao().loadSpecialDetail(id);
    }

    public void insertSpecialDetail(SpecialDetail specialDetail) {
        awakerArticleAppDatabase.specialDetailDao().insertSpecialDetail(specialDetail);
    }

    public Flowable<NewsEntity> loadNewsEntity(String id) {
        return awakerArticleAppDatabase.newsListDao().loadNewsEntity(id);
    }

    public void insertNewsEntity(NewsEntity newsEntity) {
        awakerArticleAppDatabase.newsListDao().insertNewsEntity(newsEntity);
    }

    public Flowable<CommentEntity> loadCommentEntity(String id) {
        return awakerArticleAppDatabase.commentListDao().loadCommentEntity(id);
    }

    public void insertCommentEntity(CommentEntity commentEntity) {
        awakerArticleAppDatabase.commentListDao().insertCommentEntity(commentEntity);
    }

    public Flowable<NewDetail> loadNewsDetail(String id) {
        return awakerArticleAppDatabase.newDetailDao().loadNewsDetail(id);
    }

    public void insertNewsDetail(NewDetail newDetail) {
        awakerArticleAppDatabase.newDetailDao().insertNewsDetail(newDetail);
    }

    public Flowable<List<BannerItem>> loadAllBanners() {
        return awakerArticleAppDatabase.bannerDao().loadAllBanners();
    }

    public void insertAllBanners(List<BannerItem> bannerItems) {
        awakerArticleAppDatabase.bannerDao().insertAllBanners(bannerItems);
    }

    public Flowable<SpecialListEntity> loadSpecialListEntity(String id) {
        return awakerArticleAppDatabase.specialListDao().loadSpecialListEntity(id);
    }

    public void insertAll(SpecialListEntity specialListEntity) {
        awakerArticleAppDatabase.specialListDao().insertAll(specialListEntity);
    }

    public Flowable<List<News>> loadNewsList() {
        return awakerArticleAppDatabase.newsDao().loadNewsList();
    }

    public void insertNewsList(List<News> newsList) {
        awakerArticleAppDatabase.newsDao().insertNewsList(newsList);
    }

    @Override
    public Flowable<HttpResult<List<BannerItem>>> getBanner(String token, String advType) {
        return remoteDataSource.getBanner(token, advType);
    }

    @Override
    public Flowable<HttpResult<List<News>>> getNewList(String token, int page, int id) {
        return remoteDataSource.getNewList(token, page, id);
    }

    @Override
    public Flowable<HttpResult<List<Special>>> getSpecialList(String token, int page, int cat) {
        return remoteDataSource.getSpecialList(token, page, cat);
    }

    @Override
    public Flowable<HttpResult<NewDetail>> getNewDetail(String token, String newId) {
        return remoteDataSource.getNewDetail(token, newId);
    }

    @Override
    public Flowable<HttpResult<SpecialDetail>> getSpecialDetail(String token, String id) {
        return remoteDataSource.getSpecialDetail(token, id);
    }

    @Override
    public Flowable<HttpResult<List<Comment>>> getUpNewsComments(String token, String newId) {
        return remoteDataSource.getUpNewsComments(token, newId);
    }

    @Override
    public Single<HttpResult<List<Comment>>> getNewsComments(String token, String newId,
                                                             int page) {
        return remoteDataSource.getNewsComments(token, newId, page);
    }

    @Override
    public Flowable<HttpResult<List<News>>> getHotviewNewsAll(String token, int page, int id) {
        return remoteDataSource.getHotviewNewsAll(token, page, id);
    }

    @Override
    public Flowable<HttpResult<List<News>>> getHotNewsAll(String token, int page, int id) {
        return remoteDataSource.getHotNewsAll(token, page, id);
    }

    @Override
    public Flowable<HttpResult<List<Comment>>> getHotComment(String token) {
        return remoteDataSource.getHotComment(token);
    }

    @Override
    public Flowable<HttpResult<Object>> sendNewsComment(String token, String newId, String content,
                                                        String open_id, String pid) {
        return remoteDataSource.sendNewsComment(token, newId, content, open_id, pid);
    }

    @Override
    public Flowable<HttpResult<Object>> register(String token, String email, String nickname,
                                                 String password) {
        return remoteDataSource.register(token, email, nickname, password);
    }

    @Override
    public Flowable<UserInfo> login(String token, String username, String password) {
        return remoteDataSource.login(token, username, password);
    }
}

File: article/src/main/java/com/ruzhan/awaker/article/source/AwakerRepository.kt
package com.ruzhan.awaker.article.source

import com.ruzhan.awaker.article.App
import com.ruzhan.awaker.article.db.AwakerArticleAppDatabase
import com.ruzhan.awaker.article.db.entity.CommentEntity
import com.ruzhan.awaker.article.db.entity.NewsEntity
import com.ruzhan.awaker.article.db.entity.SpecialListEntity
import com.ruzhan.awaker.article.db.entity.UserInfoEntity
import com.ruzhan.awaker.article.model.BannerItem
import com.ruzhan.awaker.article.model.Comment
import com.ruzhan.awaker.article.model.NewDetail
import com.ruzhan.awaker.article.model.News
import com.ruzhan.awaker.article.model.Special
import com.ruzhan.awaker.article.model.SpecialDetail
import com.ruzhan.awaker.article.model.UserInfo
import com.ruzhan.awaker.article.network.AwakerClient
import com.ruzhan.awaker.article.network.HttpResult
import com.ruzhan.awaker.article.source.remote.IRemoteDataSource
import com.ruzhan.awaker.article.source.remote.RemoteDataSourceImpl

import io.reactivex.Flowable
import io.reactivex.Single

class AwakerRepository private constructor() : IRemoteDataSource {

    private val remoteDataSource: IRemoteDataSource
    private val awakerArticleAppDatabase: AwakerArticleAppDatabase

    init {
        remoteDataSource = RemoteDataSourceImpl(AwakerClient.get())
        awakerArticleAppDatabase = AwakerArticleAppDatabase.get(App.get())
    }

    fun loadUserInfoEntity(id: String): Flowable<UserInfoEntity> {
        return awakerArticleAppDatabase.userInfoDao().loadUserInfoEntity(id)
    }

    fun insertUserInfoEntity(userInfoEntity: UserInfoEntity) {
        awakerArticleAppDatabase.userInfoDao().insertUserInfoEntity(userInfoEntity)
    }

    fun loadSpecialDetail(id: String): Flowable<SpecialDetail> {
        return awakerArticleAppDatabase.specialDetailDao().loadSpecialDetail(id)
    }

    fun insertSpecialDetail(specialDetail: SpecialDetail) {
        awakerArticleAppDatabase.specialDetailDao().insertSpecialDetail(specialDetail)
    }

    fun loadNewsEntity(id: String): Flowable<NewsEntity> {
        return awakerArticleAppDatabase.newsListDao().loadNewsEntity(id)
    }

    fun insertNewsEntity(newsEntity: NewsEntity) {
        awakerArticleAppDatabase.newsListDao().insertNewsEntity(newsEntity)
    }

    fun loadCommentEntity(id: String): Flowable<CommentEntity> {
        return awakerArticleAppDatabase.commentListDao().loadCommentEntity(id)
    }

    fun insertCommentEntity(commentEntity: CommentEntity) {
        awakerArticleAppDatabase.commentListDao().insertCommentEntity(commentEntity)
    }

    fun loadNewsDetail(id: String): Flowable<NewDetail> {
        return awakerArticleAppDatabase.newDetailDao().loadNewsDetail(id)
    }

    fun insertNewsDetail(newDetail: NewDetail) {
        awakerArticleAppDatabase.newDetailDao().insertNewsDetail(newDetail)
    }

    fun loadAllBanners(): Flowable<List<BannerItem>> {
        return awakerArticleAppDatabase.bannerDao().loadAllBanners()
    }

    fun insertAllBanners(bannerItems: List<BannerItem>) {
        awakerArticleAppDatabase.bannerDao().insertAllBanners(bannerItems)
    }

    fun loadSpecialListEntity(id: String): Flowable<SpecialListEntity> {
        return awakerArticleAppDatabase.specialListDao().loadSpecialListEntity(id)
    }

    fun insertAll(specialListEntity: SpecialListEntity) {
        awakerArticleAppDatabase.specialListDao().insertAll(specialListEntity)
    }

    fun loadNewsList(): Flowable<List<News>> {
        return awakerArticleAppDatabase.newsDao().loadNewsList()
    }

    fun insertNewsList(newsList: List<News>) {
        awakerArticleAppDatabase.newsDao().insertNewsList(newsList)
    }

    override fun getBanner(token: String, advType: String): Flowable<HttpResult<List<BannerItem>>> {
        return remoteDataSource.getBanner(token, advType)
    }

    override fun getNewList(token: String, page: Int, id: Int): Flowable<HttpResult<List<News>>> {
        return remoteDataSource.getNewList(token, page, id)
    }

    override fun getSpecialList(token: String, page: Int, cat: Int): Flowable<HttpResult<List<Special>>> {
        return remoteDataSource.getSpecialList(token, page, cat)
    }

    override fun getNewDetail(token: String, newId: String): Flowable<HttpResult<NewDetail>> {
        return remoteDataSource.getNewDetail(token, newId)
    }

    override fun getSpecialDetail(token: String, id: String): Flowable<HttpResult<SpecialDetail>> {
        return remoteDataSource.getSpecialDetail(token, id)
    }

    override fun getUpNewsComments(token: String, newId: String): Flowable<HttpResult<List<Comment>>> {
        return remoteDataSource.getUpNewsComments(token, newId)
    }

    override fun getNewsComments(token: String, newId: String,
                                 page: Int): Single<HttpResult<List<Comment>>> {
        return remoteDataSource.getNewsComments(token, newId, page)
    }

    override fun getHotviewNewsAll(token: String, page: Int, id: Int): Flowable<HttpResult<List<News>>> {
        return remoteDataSource.getHotviewNewsAll(token, page, id)
    }

    override fun getHotNewsAll(token: String, page: Int, id: Int): Flowable<HttpResult<List<News>>> {
        return remoteDataSource.getHotNewsAll(token, page, id)
    }

    override fun getHotComment(token: String): Flowable<HttpResult<List<Comment>>> {
        return remoteDataSource.getHotComment(token)
    }

    override fun sendNewsComment(token: String, newId: String, content: String,
                                 open_id: String, pid: String): Flowable<HttpResult<Any>> {
        return remoteDataSource.sendNewsComment(token, newId, content, open_id, pid)
    }

    override fun register(token: String, email: String, nickname: String,
                          password: String): Flowable<HttpResult<Any>> {
        return remoteDataSource.register(token, email, nickname, password)
    }

    override fun login(token: String, username: String, password: String): Flowable<UserInfo> {
        return remoteDataSource.login(token, username, password)
    }

    companion object {

        private var INSTANCE: AwakerRepository? = null

        fun get(): AwakerRepository {
            if (INSTANCE == null) {
                synchronized(AwakerRepository::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = AwakerRepository()
                    }
                }
            }
            return INSTANCE!!
        }
    }
}
